### ISSUE TITLE
Appended iframe - DOM Exception 11 error on mobile devices

### ISSUE THREAD

**Issue by ghost** [vimeo/player.js#112]
The new Vimeo API trigger a DOM Exception 11 error on mobile devices like IOS and Android on appended iframes. This issue only appears on these devices and does not trigger any error on desktop version of Chrome, Safari, IE, Firefox and Opera on Windows (7/8/10) and mac OS.
This issue also appears when a DOM manupilation was made to an iframe before to initialize the Vimeo.Player(). For example, changing the iframe src trigger this error.

Here a simple example in plain Javascript: https://jsfiddle.net/n50egh3y/1/.

This issue is extremely annoying because it's simply not possible to create web app with a load more button for example...

This issue was not present with the previous API. This issue does not appear with other video stream (like Youtube, Dailymotion, Wistia, etc...) and using the same principle.

Hope it will be solved soon...

**Comment by bdougherty** [comment]
We recently updated a manual block we had on the player end for the `play()` method on iOS because iOS requires a user gesture to start playback. As of iOS 10, it properly recognizes that user gesture when using `postMessage` to communicate with our player, which enables playback to be started via the api. We missed a check for this case though, where it's called without a user gesture. We'll try to see if we can catch this error, but you should make sure to always call `play()` from a user gesture. Otherwise, use the `autoplay=1` which will work correctly where possible.

tl;dr playback on iOS must begin from a user gesture (click or keyboard event)

**Comment by lolo34140** [comment]
Thank you for your answer.

I'm a quite a bit septic about your explanation. Because I can still reproduce this issue with a gesture like on this example: https://jsfiddle.net/pu9wLymv/

Moreover, sometimes, for an unknwon reason, the Vimeo player (with the new API) completely freeze IOS devices without any error log... i can't debug it because it requires to close Safari on IOS. It can be reproduced on my fiddles and on your live example: https://player.vimeo.com/api/demo.

**Comment by bdougherty** [comment]
There's only a limited amount of time when the user gesture applies. That play call must wait for the actual iframe itself to load and that is usually too long. The iframe needs to be loaded before `play()` is called for iOS.

Not sure what you think css styles have to do with this…

**Comment by lolo34140** [comment]
Sorry but it doesn't have any relation. Even if I load the iframe it create the same issue on IOS and Android.

Could you please take a look at the error generated by Vimeo API script: https://fiddle.jshell.net/1rL29how/
There is a real issue from your script.

P.S.: About the css isssue (it's not the case here), but on IOS, in order to handle an event like a click, the DOM which handle the gesture must have a css cursor declared (like a native HTML button) in order to be taken into account on IOS devices.

**Comment by lolo34140** [comment]
Just as a proof that's a script issue from Vimeo:

I made the same test with the attribute method mentioned in your documentation and it still trigger the same error: https://fiddle.jshell.net/ptxxvnkg/. The video load indefinitely because the script fail and stop execution due to the error : `DOM Exception 11 error`.

Hope this Vimeo issue will be fixed soon because near all ways to handle an iframe fail with new Vimeo API on IOS and Android devices.

This issue was not present with the old API and IOS 10 or Android.

**Comment by bdougherty** [comment]
That latest fiddle is still loading the player from within the click handler. In this case since the element you’re passing in is not an iframe, so the constructor makes an oEmbed call to get the iframe and then inserts it into the DOM, waits for it to load, then calls play.

The player needs to be loaded before calling play on iOS.

**Comment by lolo34140** [comment]
If you look at my previous replie,s I have provided an example with an onload event (https://fiddle.jshell.net/1rL29how/) which also trigger this error.
In the Vimeo script the video is already loaded (getOEmbedData), so there isn't any need to load it.

And if you want an example with the attribute method and a load event fully handled by Vimeo API here you go: https://fiddle.jshell.net/301a22yt/. Still the same issue.

Again, this error comes from Vimeo script. This error is not related to IOS only but also Android (and maybe other mobile devices). This error is not related to a devices but only to the Vimeo script.

**Comment by ccampbell** [comment]
@lolo34140 the problem is that you have to already have the player on the page before calling play. In both of your examples you are creating the player, appending it the DOM, and calling `play()` on it. By the time the player loads and play is called that is not considered generated by a user interaction based on whatever logic iOS/Android use.

Here is an example of it working on iOS >= 10:
https://fiddle.jshell.net/301a22yt/1/

**Comment by lolo34140** [comment]
My previous fiddles wait for the iframe with an onlad and other with the Vimeo script (XMLHttpRequest).

Your fiddle does not represent the use of case of my original question. The aim is to append an iframe (load it) and play it. A lot of websites use this principle with a load more button for example.
The new Vimeo API seems to have an error when checking the state from onlad (XMLHttpRequest) because of a lack of check on the readystate. It was not the case with the previous API. before it was possible to append and iframe, load ite then play like Youtube, Wistia, Dailymoton, etc...

**Comment by ccampbell** [comment]
@lolo34140 this didn’t work in the previous version of the API library. The play() call would have failed silently. You are correct that it didn’t throw an exception before so that is something we can address, but the actual playback would not have started. 

Can you provide a fiddle of loading the player and playing at the same time as the user gesture using YouTube’s API? Just want to see if there is something we can do to make that behavior work

**Comment by lolo34140** [comment]
Here a simple and working example with Youtube API on IOS and Android: https://fiddle.jshell.net/Ldama3j5/

There is no error, and the video autoplay when possible depending of the phone connection (cellular network of Wifi) and the bandwidth...

It's also working fine for Wistia, DailyMotion, VK and several other video stream APIs... Only Vimeo API (from the 7 video stream APIs I use) don't work on mobile devices (not only on IOS). The error comes from Vimeo script, it's not an error generated from another script as you can see in my previous fiddles.

The previous Vimeo API was working pretty fine. No errors had been triggered by the previous script. I've got more than 10 000 customers which can confirm if you want.

I hope this issue will be solved very soon. By the way, the player freeze on IOS randomly without any error. It's not a memory issue but it sound that it's due to an infinite loop...

**Comment by asha23** [comment]
Also having this issue when I have a number of players in a carousel. Oddly, lightgallery.js (with the video plugin) works perfectly fine and does exactly what it needs to. It's quite frustrating that you can't have multiple players and control them using js on iPhones. As lolo rightly says, all the other libraries (youtube, etc) work perfectly out of the box. I'd assume that Vimeo need to get this working as mobile devices are pretty important.

I have each player in the slideshow at the moment. (Cant use the dynamic embed, if you do that it doesn't work at all)

I guess I'll need to hack a workaround. My thinking is it has something to do with the play() function. This should probably wait until it receives the callback before doing anything.. It's not very good that things happen instantaneously. Is there no checking in the original library? So - If it's IOS, then do something different with the play() function? I'm no expert, only I will need to do some hacks to make this library work as it should.

I just end up with a pause button in the middle of the area and it just loads forever. I also get errors. Wasn't expecting this from an "official" library from Vimeo to be honest.

**Comment by luwes** [comment]
I believe this error is fixed in the mean time. It was caused by an uncaught promise rejection which should now be handled all correctly. Please let us know if you see otherwise. Thanks!
